security/acme-client: release 1.24

fraenki opened the pull request:
## New features
- add support for Domain-Offensive LetsEncrypt API dns_doapi (#1294)
- add support for Namecheap API (dns_namecheap)
- add support for Google Cloud DNS API dns_gcloud (#549)

## Bugfixes
- certificate status not correctly updated ( #1307)

## Enhancements
- add log message when certificate status is updated (refs #1307)

fichtner commented:
(19.1.7 has been built and is scheduled for release tomorrow so no rush here)

fichtner commented:
@fraenki I'm updating acme.sh to 2.8.1 now for eventual 19.1.8 inclusion. If you need a test package let me know.

t3nk3y commented:
I've added support for Namecheap.com which is also already supported by acme.sh

@fraenki I could clone your fork, add to it, and push in another PR.  Or, if interested in merging in my changes, they are, of course super simple:

AcmeClient.xml -> OptionValues
```xml
                        <dns_namecheap>Namecheap.com API</dns_namecheap>
```

AcmeClient.xml -> Validation
```xml
                <dns_namecheap_user type="TextField">
                    <Required>N</Required>
                </dns_namecheap_user>
                <dns_namecheap_api_key type="TextField">
                    <Required>N</Required>
                </dns_namecheap_token>
```

dialogValidation.xml
```xml
    <field>
        <label>Namecheap.com</label>
        <type>header</type>
        <style>table_dns table_dns_namecheap</style>
    </field>
    <field>
        <id>validation.dns_namecheap_user</id>
        <label>User</label>
        <type>text</type>
    </field>
    <field>
        <id>validation.dns_namecheap_api_key</id>
        <label>API Key</label>
        <type>text</type>
    </field>
```

certhelper.php
```php
            case 'dns_namecheap':
                $proc_env['NAMECHEAP_USERNAME'] = (string)$valObj->dns_nameheap_user;
                $proc_env['NAMECHEAP_API_KEY'] = (string)$valObj->dns_namecheap_api_key;
                break;
```

fraenki commented:
> @fraenki I could clone your fork, add to it, and push in another PR.

@t3nk3y Please, yes, open a new PR.

fraenki commented:
@t3nk3y I've added support for Namecheap in https://github.com/opnsense/plugins/pull/1318/commits/433d0b4e5d53bb04e314e0c1c3f2563762e11eb6. Note that is varies from your original submission, so you may need to adjust your existing config. I've removed all other comments from this PR.

fraenki commented:
Waiting for approval of https://github.com/opnsense/plugins/pull/1392.

AdSchellevis commented on security/acme-client/src/opnsense/scripts/OPNsense/AcmeClient/certhelper.php:
```diff
@@ -1203,20 +1257,62 @@ function log_cert_acme_status($certObj, $modelObj, $statusCode)
     }
 }
 
+function run_shell_command($proc_cmd, $proc_env = array())
```
@fraenki it's better to rewrite this to configd usage, if sometime in the future the webui won't have root access, this might fail (we try to avoid shell exec in our new code base as much as possible).

fraenki replied:
@AdSchellevis, thanks for the review. I like your suggestion, but there are two things to consider:

1. No command in our Let's Encrypt plugin requires root privileges. They are only run as root because this is how it works today.
1. The Google Cloud CLI requires that certain environment variables are set [1]. This is not supported by configd, AFAIK.

The latter probably prevents using configd for this purpose.

[1] https://github.com/opnsense/plugins/pull/1318/files/4132244bb050495a1f72114661f9ad7b1fe1a2a1#diff-6357d5c08a4b4115e3aea0d88e069eb1R689

AdSchellevis replied:
@fraenki our usual approach for these things is to detach both processes, so you have a program (script) that performs an action (like your gcloud setup), which uses some configuration data (usually in the form of a config/ini file or parameters if limited) and is self contained. Configd calls the action.

The main advantage of this approach is that it eases testing a lot (you can always write a unit test if needed afterwards) and ensures your code base is easier to maintain.

In this case I would probably create a configuration template to flush the settings into an ini (or json) file, which can then be picked up from your gcloud init script. Step by step your program would do the following:

1. [controller] refresh template, using configd (which writes the configuration gcloud needs)
2. [controller] exec configd action "dns_gcloud"
3. [configd] executes attached script, reports status
4. [script] read settings (from config flushed in 1)
5. [script] do actions, probably these steps: https://github.com/opnsense/plugins/blob/4132244bb050495a1f72114661f9ad7b1fe1a2a1/security/acme-client/src/opnsense/scripts/OPNsense/AcmeClient/certhelper.php#L684-L696
6. [controller] report status

The script can incorporate its own logging, can be supplied with a sample configuration file for testing purposes if needed, and could be started separately while developing/testing logic.

I'm not saying that the current code doesn't work, just noticing that maintainability could be improved and trying to prevent future disappointments (breakage after changes in core at some point in time).

fraenki replied:
> our usual approach for these things is to detach both processes

Understood, and I fully agree with your suggestion. I just want to point out that the current state of `certhelper.php` has no future anyway (at least in my opinion, it's a piece of spaghetti code that should be replaced/refactored rather sooner than later). I don't want to work on a half-baked approach that would only be implemented for _1_ out of the 47 supported DNS services.

I want to rethink the approach and rewrite `certhelper.php`, but I ask you for some more time. I don't want to start this lengthy process in the middle of implementing one of the oldest feature requests and fixing possibly the most-reported bug.

@AdSchellevis, if you agree I'd like to open a new issue for the rewrite, add your suggestions as a starting point, and merge the current state for now. Your points are all valid, but I need time to work on this.

AdSchellevis replied:
@fraenki no problem at all, I might have misread some portions as well, sometimes its best to refactor once it's more clear what the code actually should do. We all trash code from time to time :) shall I merge this now?

fraenki commented:
@fichtner please pull this in for the next 19.7 release candidate :)

fichtner commented:
Heads-up: there's not going to be another release candidate.
